### PR TITLE
MOB-2724

### PR DESCRIPTION
**Pull request by nurito-burrito**
Payments-details endpoint integration work completed.

### CODE REVIEW

File: KarhooSDK/Api/Request/Adyen/AdyenPaymentsDetailsRequest.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+public struct AdyenPaymentsDetailsRequest: Codable, KarhooCodableModel {
+    
+    public let details: AdyenPaymentDetails
+    public let paymentData: String
+    public let threeDSAuthenticationOnly: Bool
+    
+    public init(details: AdyenPaymentDetails = AdyenPaymentDetails(),
+                paymentData: String = "",
+                threeDSAuthenticationOnly: Bool = false) {
+
```
Comment (JeevanThandi) [inline KarhooSDK/Api/Request/Adyen/AdyenPaymentsDetailsRequest.swift:20]:
```suggestion
```

File: KarhooSDK/Api/Request/Adyen/AdyenPaymentsRequestPayload.swift [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+public struct AdyenPaymentsRequestPayload: Codable, KarhooCodableModel {
+    
+    public let paymentsPayload: AdyenPaymentsRequest
+    public let returnUrlSuffix: String
+
+    
+    public init(paymentsPayload: AdyenPaymentsRequest = AdyenPaymentsRequest(),
+                returnUrlSuffix: String = "") {
+
```
Comment (JeevanThandi) [inline KarhooSDK/Api/Request/Adyen/AdyenPaymentsRequestPayload.swift:19]:
```suggestion
```

File: KarhooSDK/Api/Request/Adyen/PaymentsDetailsRequestPayload.swift [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+public struct PaymentsDetailsRequestPayload: Codable, KarhooCodableModel {
+    
+    public let transactionID: String
+    public let paymentsPayload: AdyenPaymentsDetailsRequest
+
+    
+    public init(transactionID: String = "",
+                paymentsPayload: AdyenPaymentsDetailsRequest = AdyenPaymentsDetailsRequest()) {
+
```
Comment (JeevanThandi) [inline KarhooSDK/Api/Request/Adyen/PaymentsDetailsRequestPayload.swift:19]:
```suggestion
```

File: KarhooSDK/Api/Request/Adyen/PaymentsDetailsRequestPayload.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+//  Copyright © 2020 Flit Technologies Ltd. All rights reserved.
+//
+
+import Foundation
+
+public struct PaymentsDetailsRequestPayload: Codable, KarhooCodableModel {
+    
+    public let transactionID: String
+    public let paymentsPayload: AdyenPaymentsDetailsRequest
+
```
Comment (JeevanThandi) [inline KarhooSDK/Api/Request/Adyen/PaymentsDetailsRequestPayload.swift:15]:
```suggestion
```

File: KarhooSDK/Api/Response/Adyen/AdyenPayments/AdyenAction.swift [hunk 5]
```diff
@@ -47,6 +56,8 @@ public struct AdyenAction: KarhooCodableModel {
         self.data = (try? container.decode(AdyenData.self, forKey: .data)) ?? AdyenData()
         self.method = (try? container.decode(String.self, forKey: .method)) ?? ""
         self.type = (try? container.decode(String.self, forKey: .type)) ?? ""
+        self.token = (try? container.decode(String.self, forKey: .token)) ?? ""
+        self.sdkData = (try? container.decode(AdyenSDKData.self, forKey: .token)) ?? AdyenSDKData()
```
Comment (JeevanThandi) [inline KarhooSDK/Api/Response/Adyen/AdyenPayments/AdyenAction.swift:60]:
```suggestion
        self.sdkData = (try? container.decode(AdyenSDKData.self, forKey: .sdkData)) ?? AdyenSDKData()
```

File: KarhooSDKTests/TestCases/Service/Payment/AdyenPaymentsDetails/KarhooPaymentsDetailsInteractorSpec.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    /**
+     * Given: Getting Adyen payments
+     * When: Get Adyen payments request succeeds
+     * Then: Callback should be success
+     */
+    func testPaymentProviderSuccess() {
+        let mockpayment = AdyenPaymentsDetailsRequest(paymentData: "mock")
+
+        let expectedResponse = PaymentsDetailsRequestPayload(transactionID: "", paymentsPayload: mockpayment)
+        var expectedResult: Result<PaymentsDetailsRequestPayload>?
```
Comment (JeevanThandi) [inline KarhooSDKTests/TestCases/Service/Payment/AdyenPaymentsDetails/KarhooPaymentsDetailsInteractorSpec.swift:55]:
```suggestion
        var expectedResult: Result<AdyenPaymentsDetails>?
```

**Review by JeevanThandi — CHANGES_REQUESTED** [review]
couple of changes! Be sure the tests pass! 👍

File: Client/AuthRequestsTesterViewController.swift [hunk 7]
```diff
@@ -100,4 +103,39 @@ class AuthRequestsTesterViewController: UIViewController {
...
+                self?.responseLabel.text = "Success: \(result.isSuccess())"
+                self?.adyen()
+            } else {
+                self?.responseLabel.textColor = .red
+                self?.responseLabel.text = "Failed\nError: \(String(describing: result.errorValue().debugDescription))"
+            }
+        })
+    }
+    
+    private func adyen() {
```
Comment (mohadian) [inline Client/AuthRequestsTesterViewController.swift:121]: adyenPaymentdetails would be more accurate

File: Client/AuthRequestsTesterViewController.swift [hunk 8]
```diff
@@ -12,6 +12,7 @@ import KarhooSDK
 enum AuthRequestType: String {
     case login = "Login"
     case revoke = "Revoke"
+    case adyen = "Adyen"
```
Comment (mohadian) [inline Client/AuthRequestsTesterViewController.swift:15]: perhaps better to call it adyenPaymentsMethods as the adyen itself is not very self-explanatory

File: KarhooSDK/Api/Request/Adyen/AdyenPaymentDetails.swift [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+import Foundation
+
+public struct AdyenPaymentDetails: Codable, KarhooCodableModel {
+    
+    public let MD: String
+    public let PaReq: String
+    public let PaRes: String
+    public let billingToken: String
+    //public let cupsecureplus.smscode: String
```
Comment (mohadian) [inline KarhooSDK/Api/Request/Adyen/AdyenPaymentDetails.swift:17]: please remove the commented lines if they are not needed

Comment (nurito-burrito) [reply KarhooSDK/Api/Request/Adyen/AdyenPaymentDetails.swift:17]: done

File: Client/ViewController.swift [hunk 10]
```diff
@@ -13,7 +13,8 @@ class ViewController: UIViewController {
     var table: UITableView!
     let endpoints = [
         ["title": "Login", "type": "Token Exchange"],
-        ["title": "Revoke", "type": "Token Exchange"]
+        ["title": "Revoke", "type": "Token Exchange"],
+        ["title": "Adyen", "type": "Token Exchange"]
```
[the remark this answers is not in the document]
Comment (nurito-burrito) [reply Client/ViewController.swift:17]: done

File: Client/AuthRequestsTesterViewController.swift (hunk 7 above)
Comment (nurito-burrito) [reply Client/AuthRequestsTesterViewController.swift:121]: done

File: Client/AuthRequestsTesterViewController.swift (hunk 8 above)
Comment (nurito-burrito) [reply Client/AuthRequestsTesterViewController.swift:15]: done

File: Client/AppDelegate.swift [hunk 11]
```diff
@@ -14,7 +14,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     var window: UIWindow?
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-        Karhoo.set(configuration: SDKConfig())
+        Karhoo.set(configuration: AdyenSDKConfig())
```
Comment (JeevanThandi) [inline Client/AppDelegate.swift:17]:
```suggestion
        Karhoo.set(configuration: SDKConfig())
```

Comment (nurito-burrito) [reply Client/AppDelegate.swift:17]: done

**Review by JeevanThandi — APPROVED** [review]

**Review by mohadian — APPROVED** [review]

**Pull request merged**